### PR TITLE
Add document types and subtypes

### PR DESCRIPTION
**Pull request by Finemas**
Added model for document types and subtypes, update a document model for them and view at the last step.

### CODE REVIEW

File: zScanner/Services/Network/Medicalc/Model/DocumentTypeNetworkModel.swift [hunk 1]
```diff
@@ -28,7 +28,8 @@ extension DocumentTypeNetworkModel {
     func toDomainModel() -> DocumentTypeDomainModel {
         return DocumentTypeDomainModel(
             id: id,
-            name: display
+            name: display,
+            subtypes: subtypes?.map({ DocumentSubTypeDomainModel(id: $0.id, name: $0.display) }) ?? []
```
Comment (skorepak) [inline zScanner/Services/Network/Medicalc/Model/DocumentTypeNetworkModel.swift]: Use `subtypes?.map({ $0.toDomainModel })

The subtype should know how to map itself. Type should not be creating Instance of SubType

Comment (skorepak) [reply zScanner/Services/Network/Medicalc/Model/DocumentTypeNetworkModel.swift]: Should be same as in TypeDatabaseModel

File: zScanner/SupportFiles/Resources/cs.lproj/Localizable.strings [hunk 2]
```diff
@@ -67,7 +67,8 @@
 "newDocument.popAlert.cancel" = "Zrušit";
 
 // MARK: - Form
-"form.listPicker.title" = "Typ dokumentu";
+"form.listPicker.docType.title" = "Typ dokumentu";
```
Comment (skorepak) [inline zScanner/SupportFiles/Resources/cs.lproj/Localizable.strings]: Use `documentType` and `documentSubType`, do not shorten the names.

File: zScanner/ViewControllers/Documents/DocumentsListViewController.swift [hunk 3]
```diff
@@ -83,12 +83,12 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
                     self.rightBarButtons = []
                     
                 case .loading:
-                    self.rightBarButtons = [self.loadingItem]
+                    self.rightBarButtons = []
```
Comment (skorepak) [inline zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: Where is the loading now indicated?

File: zScanner/Views/Departments/DepartmentButton.swift [hunk 4]
```diff
@@ -29,6 +29,12 @@ class DepartmentButton: PrimaryButton {
     override var isSelected: Bool {
         didSet {
             backgroundColor = isSelected ? .primaryDark : .primary
+            
+            if isSelected {
```
Comment (skorepak) [inline zScanner/Views/Departments/DepartmentButton.swift]: Activity indicator have to be connected to the viewModel state, not to the selection of the button. 

We always need to react to the origin of the event, not to the side-effect. The side effect may change and the loading might break.

File: zScanner/ViewModels/Documents/DocumentsListViewModel.swift [hunk 5]
```diff
@@ -59,7 +57,7 @@ class DocumentsListViewModel {
     //MARK: Helpers
     let disposeBag = DisposeBag()
     
-    private var lastSelectedDepartment: DepartmentDomainModel?
+    var lastSelectedDepartment: DepartmentDomainModel?
```
Comment (jakubskorepapoq) [inline zScanner/ViewModels/Documents/DocumentsListViewModel.swift]: private(set)

File: zScanner/ViewControllers/Documents/DocumentsListViewController.swift [hunk 6]
```diff
@@ -83,12 +83,12 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
...
                     
                 case .success:
                     self.rightBarButtons = []
                     self.departmentsStackView.subviews.forEach({ ($0 as? UIButton)?.isSelected = false })
-                    self.coordinator.createNewDocument()
+                    self.coordinator.createNewDocument(with: self.documentsViewModel.lastSelectedDepartment!)
```
Comment (jakubskorepapoq) [inline zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: It's better to unwrap it. If you don't really have to, don't use force unwrapping. 
```
if let department = self.documentsViewModel.lastSelectedDepartment {
    self.coordinator.createNewDocument(with: department)
}
```

File: zScanner/ViewModels/Documents/DocumentsListViewModel.swift [hunk 7]
```diff
@@ -27,8 +27,6 @@ class DocumentsListViewModel {
     init(database: Database, ikemNetworkManager: NetworkManager) {
         self.database = database
         self.networkManager = ikemNetworkManager
-        
-        loadDocuments()
```
Comment (jakubskorepapoq) [inline zScanner/ViewModels/Documents/DocumentsListViewModel.swift:31]: Why did you remove it?

File: zScanner/Services/Network/Medicalc/Model/DocumentTypeNetworkModel.swift (hunk 1 above)
Comment (Finemas) [reply zScanner/Services/Network/Medicalc/Model/DocumentTypeNetworkModel.swift]: It won't work because it can't convert DocumentTypeDomainModel to DocumentSubTypeDomainModel. Should I create a subtype network model?

Comment (jakubskorepapoq) [reply zScanner/Services/Network/Medicalc/Model/DocumentTypeNetworkModel.swift]: Yes :)

File: zScanner/ViewControllers/Documents/DocumentsListViewController.swift (hunk 3 above)
Comment (Finemas) [reply zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: I moved it to the department button instead of.

File: zScanner/ViewModels/Documents/DocumentsListViewModel.swift (hunk 7 above)
Comment (Finemas) [reply zScanner/ViewModels/Documents/DocumentsListViewModel.swift:31]: It was already called in view controller with updateDocuments method so it was called twice.

File: zScanner/SupportFiles/Extensions/UIViewController.swift [hunk 8]
```diff
@@ -12,4 +12,16 @@ extension UIViewController {
...
     }
+    
+    // Whenever user gonna tap somewhere else than the current view so it will dismiss the keyboard
+    func setupKeyboardHandling() {
+        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))
+        tap.cancelsTouchesInView = false
+        view.addGestureRecognizer(tap)
+    }
+    
+    @objc func dismissKeyboard() {
```
Comment (jakubskorepapoq) [inline zScanner/SupportFiles/Extensions/UIViewController.swift]: you can use private here. this method should not be called form outside of this extension

File: zScanner/ViewControllers/Documents/DocumentsListViewController.swift [hunk 9]
```diff
@@ -83,12 +83,15 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
                     self.rightBarButtons = []
```
Comment (jakubskorepapoq) [inline zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: Please remove the `rightBarButtons` if they are not used anymore in the view controller

File: zScanner/ViewModels/Documents/DocumentsListViewModel.swift (hunk 7 above)
Comment (jakubskorepapoq) [reply zScanner/ViewModels/Documents/DocumentsListViewModel.swift:31]: I think there was some reason for it, but I don't remember it now. Let's see if it's gonna be ok :)

**Review by skorepak — APPROVED** [review]
